darts/dataprocessing/dtw/cost_matrix.py

```diff
@@ -0,0 +1,119 @@
+from typing import Tuple, Dict
+import numpy as np
+
+from .window import Window, CRWindow
+from abc import abstractmethod
+import array
+from itertools import repeat
+
+Elem = Tuple[int, int]
+
+class CostMatrix:
+    n: int
+    m: int
+
+    @abstractmethod
+    def fill(self, value: np.float): pass
+
+    @abstractmethod
+    def __getitem__(self, item): pass
+
+    @abstractmethod
+    def __setitem__(self, key, value): pass
+
+    @abstractmethod
+    def __iter__(self): pass
+
+    @abstractmethod
+    def to_dense(self) -> np.ndarray:
+        pass
+
+    @staticmethod
+    def from_window(window: Window):
+        density = len(window) / ((window.n+1)*(window.m+1))
+
+        if isinstance(window, CRWindow) and density < 0.5:
+            return SparseCostMatrix(window)
+        else:
+            return DenseCostMatrix(window.n, window.m)
+
+
+class DenseCostMatrix(np.ndarray, CostMatrix):
+    def __new__(self, n, m):
```

Comment:
I've never actually used `__new__`. Any specific reason for not using `__init__`?

Reply:
https://numpy.org/doc/stable/user/basics.subclassing.html

__new__ is called before __init__. If __new__ isn't overloaded, then ndarray __new__ function will be called, which expects a different parameter set.